File: sort/count/py/counting_sort.py
# 'k' specifies the range of key(A). [0...k].
def sort(A, k, key=lambda a: a):
    n = len(A)
    B = [0] * n
    C = [0] * (k+1)
    for i in range(n):
        C[key(A[i])] += 1
    for i in range(1,k+1):
        C[i] += C[i-1]
    for i in reversed(range(n)):
        pos = C[key(A[i])] - 1
        B[pos] = A[i]
        C[key(A[i])] -= 1
    return B

File: sort/count/java/CountingSort.java
public class CountingSort {
  // 'k' specifies the range of key(A). [0...k].
  public static int[] sort(int[] A, int k, MapInt key) {
    int[] C = new int[k+1];
    for (int i = 0; i < A.length; i++) {
      C[key.map(A[i])]++;
    }
    for (int i = 1; i < C.length; i++) {
      C[i] += C[i-1];
    }
    int[] B = new int[A.length];
    for (int i = A.length-1; i >= 0; i--) {
      int pos = C[key.map(A[i])] - 1;
      B[pos] = A[i];
      C[key.map(A[i])]--;
    }
    return B;
  }
}
